compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+        # all symbols are unreachable. Use a normal identifier and let
+        # semantic analysis figure this out
+        result = newNodeI(nkIdent, n.info)
+        result.n.ident = n[0].sym.name
+
+  else:
+    result = invalidAst(c, n)
+
+proc parseDotExpr(c: PContext, n: PNode): UntypedAst =
+  if n.len == 2:
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] =
+      case n[1].kind
+      of nkSymChoices: collapseSymChoice(c, n[1], false)
+      else:            parseIdent(c, n[1])
+  else:
+    result = invalidAstLen(c, n, 2)
+
+proc qualifiedIdent(c: PContext, n: PNode): UntypedAst =
+  case n.kind
+  of nkSym:
+    if n.sym != nil: desym(c, n.sym, n.info)
+    else:            invalidAst(c, n)
+  of nkIdent: safeIdent(c, n)
+  of nkAccQuoted: parseAccQuoted(c, n)
+  of nkOpenSymChoice, nkClosedSymChoice:
+    collapseSymChoice(c, n, true)
+  else: invalidAst(c, n)
+
+proc parseBranch(c: PContext, n: PNode, isExpr: bool): UntypedAst =
+  if n.len >= 2:
+    result = newNodeI(n.kind, n.info, n.len)
+    for i, it in sliceIt(n, 0..^2):
+      result[i] = expr(c, it)
+
+    result[^1] = if isExpr: expr(c, n[^1])
+                 else: stmt(c, n[^1])
+  else:
+    result = invalidAst(c, n)
+
+proc process*(c: PContext, n: PNode): UntypedAst =
+  template checkMinSonsLen(n: PNode, length: int) =
+    if n.len < length:
+      {.line.}:
+        return toUntyped c.config.newError(n, SemReport(kind: rsemIllformedAst, ast: n,
+          str: "Expected at least $1 elements, but found $2" % [$length, $n.len]))
+
+  template checkSonsLen(n: PNode, length: int) =
+    if n.len != length:
+      {.line.}:
+        return toUntyped c.config.newError(n, SemReport(kind: rsemIllformedAst, ast: n,
+          str: "Expected $1 elements, but found $2" % [$length, $n.len]))
+
+  template invalid() =
+    result = toUntyped c.config.newError(n, SemReport(kind: rsemIllformedAst, ast: n))
+
+  case n.kind
+  of nkNone:
+    invalid()
+  of nkEmpty:
+    result = newNodeI(nkEmpty, n.info)
+  of nkIdent:
+    if n.ident != nil:
+      result = newIdentNode(n.ident, n.info)
+    else:
+      invalid()
+  of nkSym:
+    if n.sym != nil:
+      result = desym(c, n.sym, n.info)
+    else:
+      invalid()
+  of nkType:
+    # try to turn it into an identifier
+    if n.typ != nil and n.typ.sym != nil:
+      result = desym(c, n.sym, n.info)
+    else:
+      invalid()
+  of nkCharLit..nkUInt64Lit:
+    result = block:
+      let r = ast.newIntNode(n.kind, n.intVal)
+      r.info = n.info
+      toUntyped r
+  of nkFloatLit..nkFloat128Lit:
+    result = block:
+      let r = ast.newNodeI(n.kind, n.info)
+      r.floatVal = n.floatVal
+      toUntyped r
+  of nkStrLit..nkTripleStrLit:
+    result = block:
+      let r = ast.newNodeI(n.kind, n.info)
+      r.strVal = n.strVal
+      toUntyped r
+  of nkNilLit:
+    result = newNodeI(nkNilLit, n.info)
+  of nkDotCall:
+    # XXX: hm, turn into an ``nkDotExpr`` + ``nkCall`` instead?
```

Comment:
Yes, I think that would be a more true unwind.

Reply:
The reason I hesitated here is because `nkDotCall` can appear in neither typed nor untyped AST a macro gets from sem. Looking at `nkDotCall` a bit closer, I think the more proper unwind would be also be only `nkDotExpr`. That is:

```
DotCall (x, y) -> DotExpr (y, x)
```

Reply:
Good catch, that's indeed a truer reversal based on what `dotTransformation` does.